File: 42. Trapping Rain Water/Trapping Rain Water.cpp
class Solution {
public:
    int trap(vector<int>& h) {
        int l = 0;
        int r = h.size() - 1;
        int maxL = h[0];
        int maxR = h[r];
        int result = 0;

        while (l < r) {
            if (maxL < maxR) {
                l++;
                maxL = max(maxL, h[l]);
                result += (maxL - h[l]);
            } else {
                r--;
                maxR = max(maxR, h[r]);
                result += (maxR - h[r]);
            }
        }

        return result;
    }
};

File: 42. Trapping Rain Water/Trapping Rain Water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function (height) {
  let l = 0;
  let r = height.length - 1;
  let maxL = height[0];
  let maxR = height[r];
  let result = 0;

  while (l < r) {
    if (maxL < maxR) {
      l++;
      maxL = Math.max(maxL, height[l]);
      result += maxL - height[l];
    } else {
      r--;
      maxR = Math.max(maxR, height[r]);
      result += maxR - height[r];
    }
  }

  return result;
};
